### PR TITLE
[ML] Assorted time series modelling improvements

### PR DESCRIPTION
**Pull request by tveasey**
This merges a long running branch which:
1. Improves modularity, robustness and testing of code to select seasonal components to model.
2. Improves re-initialisation of seasonal components after detecting new or different components.
3. Reworks and unifies the code to detect change points in time series and extend to detect new types of change, such as a broader range of  time shifts.
4. Improves handling of extended anomalous regions which subsequently revert to normal behaviour (things like bank holidays) that should have minimal impact on the model.
5. Improves handling of very long bucket lengths.

### CODE REVIEW

File: lib/api/CAnomalyJob.cc [hunk 1]
```diff
@@ -90,8 +90,8 @@ const std::string INTERIM_BUCKET_CORRECTOR_TAG("k");
 //! versions cannot read the model state of newer versions following a breaking
 //! change.  This constant tells the node assignment code not to load new model states
 //! on old nodes in a mixed version cluster.  (The last breaking change was in 7.9 in
```
Comment (droberts195) [inline lib/api/CAnomalyJob.cc:92]:
```suggestion
//! on old nodes in a mixed version cluster.  (The last breaking change was in 7.11 in
```

File: lib/api/CAnomalyJob.cc [hunk 2]
```diff
@@ -90,8 +90,8 @@ const std::string INTERIM_BUCKET_CORRECTOR_TAG("k");
...
-//! lib/core/CPackedBitVector.cc in https://github.com/elastic/ml-cpp/pull/1340.)
-const std::string MODEL_SNAPSHOT_MIN_VERSION("7.9.0");
+//! lib/core/CPackedBitVector.cc in https://github.com/elastic/ml-cpp/pull/1550.)
```
Comment (droberts195) [inline lib/api/CAnomalyJob.cc:93]:
```suggestion
//! lib/maths/CTimeSeriesDecomposition.cc in https://github.com/elastic/ml-cpp/pull/1614.)
```

**Review by droberts195 — APPROVED** [review]
I haven't looked through the changes in this PR but they all should have been reviewed previously on the feature branch so I am happy to merge once CI is green and QA are ready.

**Comment by tveasey** [comment]
retest

**Pull request merged**